File: python/aim_node/gateway_v2/connect.py
from __future__ import annotations

from dataclasses import asdict, dataclass
from typing import Any

import httpx

from .contracts import CLIENT_METHODS, GatewaySurface, reject_payload_contract_fields


CONNECT_CLIENT_METHOD = CLIENT_METHODS[GatewaySurface.CONNECT]
LOCAL_SECRET_REF_PREFIX = "local://aim-node/secrets/"


@dataclass(frozen=True)
class ConnectorPreferences:
    preferred_connector_ids: list[str] | None = None
    preferred_seller_edge_id: str | None = None
    connection_types: list[str] | None = None
    region: str | None = None


@dataclass(frozen=True)
class ConnectRequest:
    metadata: dict[str, Any]
    quote_id: str
    billing_session_id: str
    delivery_mode: str
    accepted_terms_hash: str
    connector_preferences: ConnectorPreferences | None = None
    seller_edge_id: str | None = None


@dataclass(frozen=True)
class SellerEdgeBinding:
    seller_edge_id: str
    seller_id: str
    connector_id: str
    listing_id: str
    identity_hash: str


@dataclass(frozen=True)
class GrantConnection:
    type: str
    local_secret_ref: str
    seller_edge: SellerEdgeBinding | None = None


@dataclass(frozen=True)
class AllowedScope:
    listing_id: str
    quote_id: str
    billing_session_id: str
    delivery_mode: str
    operations: list[str]
    max_concurrency: int


@dataclass(frozen=True)
class ConnectResponse:
    access_grant_id: str
    connector_id: str
    state: str
    expires_at: str
    connection: GrantConnection
    allowed_scope: AllowedScope
    terms_hash: str


class GatewayConnectClient:
    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        client: httpx.Client | None = None,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.client = client or httpx.Client()

    def connect(self, request: ConnectRequest) -> ConnectResponse:
        reject_payload_contract_fields(list(asdict(request).keys()))
        _reject_raw_seller_secret_fields(asdict(request))
        headers = {
            "content-type": "application/json",
            "x-request-id": str(_pick(request.metadata, "requestId", "request_id")),
        }
        idempotency_key = _pick(request.metadata, "idempotencyKey", "idempotency_key")
        if idempotency_key:
            headers["idempotency-key"] = str(idempotency_key)
        if self.api_key:
            headers["authorization"] = f"Bearer {self.api_key}"

        response = self.client.post(
            f"{self.base_url}/v1/gateway/connect",
            json=_drop_none(asdict(request)),
            headers=headers,
        )
        response.raise_for_status()
        return parse_connect_response(response.json())


def local_grant_secret_ref(access_grant_id: str) -> str:
    if not access_grant_id:
        raise ValueError("access_grant_id is required")
    return f"{LOCAL_SECRET_REF_PREFIX}grants/{access_grant_id}"


def parse_connect_response(data: dict[str, Any]) -> ConnectResponse:
    connection = _pick(data, "connection")
    allowed_scope = _pick(data, "allowed_scope", "allowedScope")
    secret_ref = _pick(connection, "local_secret_ref", "localSecretRef")
    _assert_local_secret_ref(secret_ref)
    return ConnectResponse(
        access_grant_id=_pick(data, "access_grant_id", "accessGrantId"),
        connector_id=_pick(data, "connector_id", "connectorId"),
        state=data["state"],
        expires_at=_pick(data, "expires_at", "expiresAt"),
        connection=GrantConnection(
            type=connection["type"],
            local_secret_ref=secret_ref,
            seller_edge=_parse_seller_edge(_pick(connection, "seller_edge", "sellerEdge")),
        ),
        allowed_scope=AllowedScope(
            listing_id=_pick(allowed_scope, "listing_id", "listingId"),
            quote_id=_pick(allowed_scope, "quote_id", "quoteId"),
            billing_session_id=_pick(allowed_scope, "billing_session_id", "billingSessionId"),
            delivery_mode=_pick(allowed_scope, "delivery_mode", "deliveryMode"),
            operations=list(allowed_scope["operations"]),
            max_concurrency=_pick(allowed_scope, "max_concurrency", "maxConcurrency"),
        ),
        terms_hash=_pick(data, "terms_hash", "termsHash"),
    )


def _parse_seller_edge(data: dict[str, Any] | None) -> SellerEdgeBinding | None:
    if not data:
        return None
    return SellerEdgeBinding(
        seller_edge_id=_pick(data, "seller_edge_id", "sellerEdgeId"),
        seller_id=_pick(data, "seller_id", "sellerId"),
        connector_id=_pick(data, "connector_id", "connectorId"),
        listing_id=_pick(data, "listing_id", "listingId"),
        identity_hash=_pick(data, "identity_hash", "identityHash"),
    )


def _assert_local_secret_ref(secret_ref: str) -> None:
    if not isinstance(secret_ref, str) or not secret_ref.startswith(LOCAL_SECRET_REF_PREFIX):
        raise ValueError("secret_ref must point to the local aim-node secret store")


def _reject_raw_seller_secret_fields(data: dict[str, Any]) -> None:
    for field_name in data:
        lowered = field_name.lower()
        if "raw_secret" in lowered or "seller_secret" in lowered:
            raise ValueError("raw seller secrets are forbidden in gateway.connect")


def _pick(data: dict[str, Any], *keys: str) -> Any:
    for key in keys:
        if key in data:
            return data[key]
    return None


def _drop_none(data: dict[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in data.items() if value is not None}

File: src/gateway_v2/connect.ts
import { assertNoPayloadFields, clientMethods } from "./client_contracts";
import type { OpaqueId, SharedRequestMetadata } from "./client_contracts";
import { assertLocalSecretRef, assertNoRawSellerSecrets, type LocalSecretRef } from "./local_secret_refs";

export const connectClientMethod = clientMethods.connect;

export interface ConnectorPreferences {
  preferredConnectorIds?: OpaqueId[];
  preferredSellerEdgeId?: OpaqueId;
  connectionTypes?: string[];
  region?: string;
}

export interface ConnectRequest {
  metadata: SharedRequestMetadata;
  quoteId: OpaqueId;
  billingSessionId: OpaqueId;
  deliveryMode: string;
  connectorPreferences?: ConnectorPreferences;
  acceptedTermsHash: string;
  sellerEdgeId?: OpaqueId;
}

export interface SellerEdgeBinding {
  sellerEdgeId: OpaqueId;
  sellerId: OpaqueId;
  connectorId: OpaqueId;
  listingId: OpaqueId;
  identityHash: string;
}

export interface GrantConnection {
  type: "local" | "seller_edge";
  localSecretRef: LocalSecretRef;
  sellerEdge?: SellerEdgeBinding;
}

export interface AllowedScope {
  listingId: OpaqueId;
  quoteId: OpaqueId;
  billingSessionId: OpaqueId;
  deliveryMode: string;
  operations: string[];
  maxConcurrency: number;
}

export interface ConnectResponse {
  accessGrantId: OpaqueId;
  connectorId: OpaqueId;
  state: "active" | "revoked";
  expiresAt: string;
  connection: GrantConnection;
  allowedScope: AllowedScope;
  termsHash: string;
}

export interface GatewayConnectClientOptions {
  baseUrl: string;
  apiKey?: string;
  fetch?: typeof fetch;
}

export class GatewayConnectClient {
  private readonly baseUrl: string;
  private readonly apiKey?: string;
  private readonly fetchImpl: typeof fetch;

  constructor(options: GatewayConnectClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.apiKey = options.apiKey;
    this.fetchImpl = options.fetch ?? fetch;
  }

  async connect(request: ConnectRequest): Promise<ConnectResponse> {
    assertNoPayloadFields(Object.keys(request));
    assertNoRawSellerSecrets(request as unknown as Record<string, unknown>);

    const headers: Record<string, string> = {
      "content-type": "application/json",
      "x-request-id": request.metadata.requestId,
    };
    if (request.metadata.idempotencyKey) {
      headers["idempotency-key"] = request.metadata.idempotencyKey;
    }
    if (this.apiKey) {
      headers.authorization = `Bearer ${this.apiKey}`;
    }

    const response = await this.fetchImpl(`${this.baseUrl}/v1/gateway/connect`, {
      method: "POST",
      headers,
      body: JSON.stringify(request),
    });
    if (!response.ok) {
      throw new Error(`gateway connect failed: ${response.status} ${response.statusText}`);
    }
    return parseConnectResponse(await response.json());
  }
}

export function parseConnectResponse(data: Record<string, unknown>): ConnectResponse {
  const connection = pickRecord(data, "connection");
  const allowedScope = pickRecord(data, "allowed_scope", "allowedScope");
  const localSecretRef = pickString(connection, "local_secret_ref", "localSecretRef");
  assertLocalSecretRef(localSecretRef);

  return {
    accessGrantId: pickString(data, "access_grant_id", "accessGrantId"),
    connectorId: pickString(data, "connector_id", "connectorId"),
    state: pickString(data, "state") as ConnectResponse["state"],
    expiresAt: pickString(data, "expires_at", "expiresAt"),
    connection: {
      type: pickString(connection, "type") as GrantConnection["type"],
      localSecretRef,
      sellerEdge: parseSellerEdge(pickOptionalRecord(connection, "seller_edge", "sellerEdge")),
    },
    allowedScope: {
      listingId: pickString(allowedScope, "listing_id", "listingId"),
      quoteId: pickString(allowedScope, "quote_id", "quoteId"),
      billingSessionId: pickString(allowedScope, "billing_session_id", "billingSessionId"),
      deliveryMode: pickString(allowedScope, "delivery_mode", "deliveryMode"),
      operations: pickStringArray(allowedScope, "operations"),
      maxConcurrency: pickNumber(allowedScope, "max_concurrency", "maxConcurrency"),
    },
    termsHash: pickString(data, "terms_hash", "termsHash"),
  };
}

function parseSellerEdge(data?: Record<string, unknown>): SellerEdgeBinding | undefined {
  if (!data) {
    return undefined;
  }
  return {
    sellerEdgeId: pickString(data, "seller_edge_id", "sellerEdgeId"),
    sellerId: pickString(data, "seller_id", "sellerId"),
    connectorId: pickString(data, "connector_id", "connectorId"),
    listingId: pickString(data, "listing_id", "listingId"),
    identityHash: pickString(data, "identity_hash", "identityHash"),
  };
}

function pickRecord(data: Record<string, unknown>, ...keys: string[]): Record<string, unknown> {
  const value = pick(data, ...keys);
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`expected object field ${keys[0]}`);
  }
  return value as Record<string, unknown>;
}

function pickOptionalRecord(data: Record<string, unknown>, ...keys: string[]): Record<string, unknown> | undefined {
  const value = pick(data, ...keys);
  if (value === undefined || value === null) {
    return undefined;
  }
  if (typeof value !== "object" || Array.isArray(value)) {
    throw new Error(`expected object field ${keys[0]}`);
  }
  return value as Record<string, unknown>;
}

function pickString(data: Record<string, unknown>, ...keys: string[]): string {
  const value = pick(data, ...keys);
  if (typeof value !== "string" || value.length === 0) {
    throw new Error(`expected string field ${keys[0]}`);
  }
  return value;
}

function pickNumber(data: Record<string, unknown>, ...keys: string[]): number {
  const value = pick(data, ...keys);
  if (typeof value !== "number") {
    throw new Error(`expected number field ${keys[0]}`);
  }
  return value;
}

function pickStringArray(data: Record<string, unknown>, ...keys: string[]): string[] {
  const value = pick(data, ...keys);
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string")) {
    throw new Error(`expected string array field ${keys[0]}`);
  }
  return value;
}

function pick(data: Record<string, unknown>, ...keys: string[]): unknown {
  for (const key of keys) {
    if (key in data) {
      return data[key];
    }
  }
  return undefined;
}
